### PR TITLE
Only apply transformGroupedWindows() with ORCA.

### PR DESCRIPTION
**Pull request by hlinnaka**
The Postgres planner doesn't need it. Move the code to do it, so that it's
only used before passing a tree to ORCA. This doesn't change anything with
ORCA, but with the Postgres planner, it has some benefits:

* Some cases before this patch do not give correct results and now they run
  correctly (e.g. case `regress/olap_window_seq`)
* Fixes github issue #10143.

### CODE REVIEW

**Comment by vraghavan78** [comment]
@asubramanya it should not be ported to 6X

**Comment by hlinnaka** [comment]
On 17/06/2020 04:27, Hans Zeller wrote:
> *@zellerh* commented on this pull request.
> 
> Overall, I like the change. We could also try to move this code into the 
> algebrizer.

That would be great. That's well beyond my skills, however.

> Related to that, here is one question, though: Wouldn't we 
> have to call |transformGroupedWindows| for every |Query|, not just for 
> the top one?

Yes, we definitely should recurse into sub-Querys too.
Good catch.

- Heikki

**Comment by hlinnaka** [comment]
On 17/06/2020 20:33, Venkatesh Raghavan wrote:
> @asubramanya it should not be ported to 6X

Any particular reason? It does fix an existing bug or two. I'm not sure 
what our usual policy for backporting is, but 6X is definitely affected.

- Heikki

**Comment by zellerh** [comment]
@asubramanya and I can look into moving the code into the algebrizer.

**Comment by vraghavan78** [comment]
> On 17/06/2020 20:33, Venkatesh Raghavan wrote: @asubramanya it should not be ported to 6X
> Any particular reason? It does fix an existing bug or two. I'm not sure what our usual policy for backporting is, but 6X is definitely affected. - Heikki

Let me rephrase. I am not against it eventually going to 6x. Just now right aways.

Incidentally @zellerh @asubramanya are working on a similar area as we speak. We have one PR https://github.com/greenplum-db/gpdb/pull/10309 in PR already. @asubramanya is working on the same area. I would like these three fixes to go together on 6 to minimize risk. Master, these fixes can land independently. @zellerh or @asubramanya or orca team can take the responsibility of doing that.

**Comment by kainwen** [comment]
> On 17/06/2020 20:33, Venkatesh Raghavan wrote: @asubramanya it should not be ported to 6X
> Any particular reason? It does fix an existing bug or two. I'm not sure what our usual policy for backporting is, but 6X is definitely affected. - Heikki

There is a jira that reported for 6x and is just caused by this bug. So we need to backport to 6X.

File: src/backend/optimizer/plan/orca.c [hunk 1]
```diff
@@ -970,6 +991,11 @@ grouped_window_mutator(Node *node, void *context)
 					 errhint("You might need to use explicit aliases and/or to refer to grouping keys in the same way throughout the query, or turn optimizer=off.")));
 		}
 	}
+	else if (IsA(node, SubLink))
+	{
+		/* put the subquery into Q'' */
+		result = (Node *) var_for_gw_expr(ctx, node, true);
```
Comment (chrishajas) [inline src/backend/optimizer/plan/orca.c:997]: Can we rename gw to grouped_window (looks like that's what we use for the rest of the references). Also, was there a test case that exposed this?

Comment (hardikar) [inline src/backend/optimizer/plan/orca.c:997]:
```suggestion
		result = (Node *) var_for_gw_expr(ctx, node, true /* force */);
```

File: src/test/regress/expected/bfv_olap.out [hunk 2]
```diff
@@ -645,6 +646,46 @@ group by a.code;
...
+select * from (select sum(a.salary) over(), count(*)
+               from t2_github_issue_10143 a
+               group by a.salary) T;
+ sum  | count 
+------+-------
+ 2100 |     1
+ 2100 |     1
+(2 rows)
+
+-- this query currently falls back, needs to be fixed
```
Comment (hardikar) [inline src/test/regress/expected/bfv_olap.out:658]: Are we tracking this in issues/tracker? Also I *think* we use FIXME's but it doesn't matter, any way to make this easier to find the future, I guess, would be great.

Comment (zellerh) [reply src/test/regress/expected/bfv_olap.out:658]: Yes, we added a story on our team-internal tracking tool.

File: src/backend/optimizer/plan/orca.c (hunk 1 above)
Comment (zellerh) [reply src/backend/optimizer/plan/orca.c:997]: Done.

File: src/backend/optimizer/plan/orca.c [hunk 3]
```diff
@@ -400,3 +411,761 @@ push_down_expr_mutator(Node *node, List *child_tlist)
...
+		rte = makeNode(RangeTblEntry);
+		rte->rtekind = RTE_SUBQUERY;
+		rte->subquery = subq;
+		rte->alias = NULL;			/* fill in later */
+		rte->eref = NULL;			/* fill in later */
+		rte->inFromCl = true;
+		rte->requiredPerms = ACL_SELECT;
+
+		/*
+		 * Default? rte->inh = 0; rte->checkAsUser = 0;
```
Comment (ashuka24) [inline src/backend/optimizer/plan/orca.c:568]: could you clarify this comment? why is this question>

File: src/backend/optimizer/plan/orca.c [hunk 4]
```diff
@@ -400,3 +411,761 @@ push_down_expr_mutator(Node *node, List *child_tlist)
...
+ * Prime the subquery target list in the context with the grouping
+ * and windowing attributes from the given query and adjust the
+ * subquery group clauses in the context to agree.
+ *
+ * Note that we arrange dense sortgroupref values and stash the
+ * referents on the front of the subquery target list.  This may
+ * be over-kill, but the grouping extension code seems to like it
+ * this way.
+ *
+ * Note that we only transfer sortgrpref values associated with
```
Comment (ashuka24) [inline src/backend/optimizer/plan/orca.c:665]:
```suggestion
* Note that we only transfer sortgroupref values associated with
```

File: src/backend/optimizer/plan/orca.c [hunk 5]
```diff
@@ -400,3 +411,761 @@ push_down_expr_mutator(Node *node, List *child_tlist)
...
+		var->location = 0;
+	}
+
+	return var;
+}
+
+
+/* Helper for transformGroupedWindows:
+ *
+ * Mutator for subquery groupingClause to adjust sortgrpref values
```
Comment (ashuka24) [inline src/backend/optimizer/plan/orca.c:799]:
```suggestion
 * Mutator for subquery groupingClause to adjust sortgroupref values
```

File: src/backend/optimizer/plan/orca.c [hunk 6]
```diff
@@ -400,3 +411,761 @@ push_down_expr_mutator(Node *node, List *child_tlist)
...
+		}
+		new_tle->ressortgroupref = tle->ressortgroupref;
+		new_tle->resorigtbl = InvalidOid;
+		new_tle->resorigcol = 0;
+		new_tle->resjunk = tle->resjunk;
+
+		/*
+		 * This is pretty shady, but we know our call pattern.  The target
+		 * list is at level 1, so we're interested in target entries at level
+		 * 2.  We record them in context so var_for_gw_expr can maybe make a
```
Comment (ashuka24) [inline src/backend/optimizer/plan/orca.c:930]:
``` suggestion
		 * 2.  We record them in context so var_for_grouped_window_expr can maybe make a
```

File: src/backend/optimizer/plan/orca.c [hunk 7]
```diff
@@ -400,3 +411,761 @@ push_down_expr_mutator(Node *node, List *child_tlist)
...
+
+	inrte = rt_fetch(invar->varno, rtable);
+
+	if (inrte->rtekind == RTE_JOIN)
+	{
+		Node	   *ja = list_nth(inrte->joinaliasvars, invar->varattno - 1);
+
+		/*
+		 * Though Node types other than Var (e.g., CoalesceExpr or Const) may
+		 * occur as joinaliasvars, we ignore them.
```
Comment (ashuka24) [inline src/backend/optimizer/plan/orca.c:1120]: Why do we ignore other node types?

File: src/backend/optimizer/plan/orca.c (hunk 3 above)
Comment (zellerh) [reply src/backend/optimizer/plan/orca.c:568]: I think the MakeNode macro initializes all the fields to 0 with memset, so this is probably just a reminder that we leave those fields at 0. Note that this is old code that just got moved by this PR.

File: src/backend/optimizer/plan/orca.c (hunk 7 above)
Comment (zellerh) [reply src/backend/optimizer/plan/orca.c:1120]: Sorry, I don't have a good answer for that. This code just got moved and I didn't look at all of its details. In this context we are just looking for `Var`s, so I assume that's why we ignore anything else.

File: src/backend/optimizer/plan/orca.c (hunk 5 above)
Comment (zellerh) [reply src/backend/optimizer/plan/orca.c:799]: Done.

File: src/backend/optimizer/plan/orca.c (hunk 4 above)
Comment (zellerh) [reply src/backend/optimizer/plan/orca.c:665]: Done.

**Comment by zellerh** [comment]
@kainwen, we might have to come up with a solution for 6X. @asubramanya tells me that this PR causes test failures when it's just ported to 6X. What I can try is just port the bug fix we added and see what that does. Looking at issue https://github.com/greenplum-db/gpdb/issues/10143 I'm reasonably confident that will fix the problem.

**Comment by zellerh** [comment]
Created https://github.com/greenplum-db/gpdb/pull/10366.
